### PR TITLE
kvserver: deflake lease preferences during outage

### LINKED ISSUE #105101: kv/kvserver: TestLeasePreferencesDuringOutage failed
kv/kvserver.TestLeasePreferencesDuringOutage [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10573809?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/10573809?buildTab=artifacts#/) on release-22.2 @ [8055184fc991424d1d0eefd7ccf703948b8de3ee](https://github.com/cockroachdb/cockroach/commits/8055184fc991424d1d0eefd7ccf703948b8de3ee):

```
=== RUN   TestLeasePreferencesDuringOutage
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestLeasePreferencesDuringOutage4182034325
    test_log_scope.go:79: use -show-logs to present logs inline
    client_lease_test.go:1028: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_lease_test.go:1028
        	Error:      	Received unexpected error:
        	            	replica [n1,s1,r53/1:/{Table/100-Max}] does not have the range lease: [NotLeaseHolderError] lease acquisition attempt lost to another lease, which has expired in the meantime; r53: replica (n1,s1):1 not lease holder; lease holder unknown
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Enqueue
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store.go:3644
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestLeasePreferencesDuringOutage
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver_test/pkg/kv/kvserver/client_lease_test.go:1026
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1446
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (2) replica [n1,s1,r53/1:/{Table/100-Max}] does not have the range lease
        	            	Wraps: (3) [NotLeaseHolderError] lease acquisition attempt lost to another lease, which has expired in the meantime; r53: replica (n1,s1):1 not lease holder; lease holder unknown
        	            	Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *roachpb.NotLeaseHolderError
        	Test:       	TestLeasePreferencesDuringOutage
        	Messages:   	failed to enqueue replica for replication
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestLeasePreferencesDuringOutage4182034325
--- FAIL: TestLeasePreferencesDuringOutage (6.38s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeasePreferencesDuringOutage.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-28862

### PR DESCRIPTION
**Pull request by kvoli**
It was possible that another node would acquire the lease, causing the replicate queue enqueue to fail. Wrap the enqueue in a succeeds soon to retry when this happens.

Note this doesn't unskip the test on the master branch, this is targeted at the 22.2 release branch.

Resolves: #105101

Release note: None

### CODE REVIEW

File: pkg/kv/kvserver/client_lease_test.go [hunk 1]
```diff
@@ -1035,10 +1035,16 @@ func TestLeasePreferencesDuringOutage(t *testing.T) {
...
-	_, _, enqueueError := tc.GetFirstStoreFromServer(t, 0).
-		Enqueue(ctx, "replicate", repl, true /* skipShouldQueue */, false /* async */)
 
-	require.NoError(t, enqueueError, "failed to enqueue replica for replication")
+	testutils.SucceedsSoon(t, func() error {
+		// n1 requires the lease to process the replica through the replicate
+		// queue. It will try and acquire the lease if it doesn't have it. It is
+		// possible that another node gets there before n1 - expect the acquisition
+		// to succeed eventually.
+		_, _, enqueueError := tc.GetFirstStoreFromServer(t, 0).
```
Comment (aliher1911) [inline pkg/kv/kvserver/client_lease_test.go:1044]: Will this ever succeed? If we lost lease to some other store in general case we try redirectOnOrAcquireLease() which will never steal the lease if it is valid. It is not bypassed by skipShouldQueue.

If that is correct, we should enqueue on the store where the lease is if we got a lease error.

**Review by aliher1911 — CHANGES_REQUESTED** [review]
Can you verify that this is not the case?

File: pkg/kv/kvserver/client_lease_test.go (hunk 1 above)
Comment (kvoli) [reply pkg/kv/kvserver/client_lease_test.go:1044]: In the failures I've reproduced, the lease is invalid. The node which beats n1 to the lease is artificially dead. 

```
replica [n1,s1,r53/1:/{Table/100-Max}] does not have the range lease: [NotLeaseHolderError] lease acquisition attempt lost to another lease, which has expired in the meantime; r53: replica (n1,s1):1 not lease holder; lease holder unknown
```


I think you're right, though. If it happened to move to one of the live replicas in the range (n4) then we'd need to enqueue there instead.

With this patch I've stressed:

```
11579 runs so far, 0 failures, over 52m30s
```

vs without

```
3928 runs completed, 1 failures, over 17m33s
```

I'll continue to let the patch stress overnight. I

**Comment by kvoli** [comment]
Closing this PR, I'm going to revisit this approach for 22.2.

**Pull request closed without merging**